Customer 라우팅/필터링 추가

handsone-u opened the pull request:
### StripPrefix

> url에서 2층을 떼어내고 다운스트림 서비스로 전달.


### PreAuthorizationFilterFactory

> token을 검증하고 헤더에 인증 정보를 추가하는 역할.


- request를 보고 로그인과 가입 요청일 시 다음 필터로 넘김. `chain.filter(exchange);`
- 검증이 성공할 시 응답헤더 조작. 
  ```
  chain.filter(exchange.mutate()
                .request(mutatedRequest)
                .build());
  ```


### PostAuthenticationFilterFactory:

> 로그인 요청 후 API 응답으로 부터 JWT 생성하고 응답 헤더에 추가하는 역할.


ModifyResponseBodyGatewayFilterFactory: 응답 body 읽고/수정.
위 필터는 라우팅이 완료된 다음에 동작함.
- 가입 요청일 시 다음 필터로 넘김.
- response로 부터 받은 응답 결과로 JWT 만들어 헤더에 추가함.

### 필터 설정

> 필터 순서대로 동작, 필터는 다운 스트림 서비스로의 라우팅 전/후 모두 처리함.


  ```
        filters:
            - StripPrefix=2
            - name: PreAuthorizationFilterFactory
              args:
                role: CUSTOMER
            - name: PostAuthenticationFilterFactory
              args:
                role: CUSTOMER
```

### 요청 흐름

로그인 요청 -> **StripPrefix** -> **PreAuthorizationFilter**. 로그인 url 분기 -> **PostAuthenticationFilter**.라우팅 & 다운 스트림 서비스에서의 로그인 응답으로 JWT 만들어 헤더에 추가 -> Client에게 응답

일반 요청 -> **StripPrefix** -> **PreAuthorizationFilter**. jwt 검증 및 헤더에 Id, Role 값 추가. -> 다운 스트림 서비스로 라우팅 ...

---

### 인증 정보. 헤더를 통해 다운 스트림 서비스로 
```
          ServerHttpRequest mutatedRequest = exchange.getRequest().mutate()
              .header("X-User-Id", provider.extractId(token).toString())
              .header("X-User-Role", provider.extractRole(token))
              .build();

```
다운 스트림 서비스에서 Spring Security 와 인증 필터를 추가해 사용자의 인증 정보를 HTTP 헤더를 통해 받을 수 있도록 구현했습니다.

handsone-u committed:
[#4] add: 라우팅 추가

handsone-u committed:
[#4] feat: jwt인증/인가 필터 구현

handsone-u committed:
[#4] imporve: 토큰 검증 시 필요로 하는 role을 필터에서 설정으로 관리

- 필터의 중복 구현 없이 환경설정을 통해 주입한 args로 token의 role을 검증하고 생성할 수 있도록 구현.

JinTaekLim commented on src/main/resources/application-local.yml:
현재는 ROLE 값을 기준으로 각 도메인에 접근 권한을 설정하고 있는 걸로 이해했습니다.

다만 이렇게 지정하게 되면 크게 두 가지 문제점이 있을 것같아요.
1. 게이트웨이에서 모든 도메인 API 의 접근 권한을 알아야함 
2. 권한이 없어도 접근할 수 있는 경로에 대한 처리를 별도로 처리해야함


때문에, 게이트웨이에서는 JWT 토큰 정보 유무 확인 후 존재할시 헤더에 id 값을 넣고, 존재하지 않으면 별도의 처리를 하지 않는 방향으로 처리하는 건 어떨까 싶습니다.
도메인 내부에서 접근 권한 처리를 할 수 있도록요!

JinTaekLim commented on src/main/java/com/baedal/gateway/filter/JwtFilterConfig.java:
고정된 값들을 관리하기 위한 파일로 이해했습니다.
그렇다면 @Data 가 아니라 @Getter 만으로도 충분하지 않을까요?

handsone-u replied:
yml 에서 args로 가져오는 설정 값인데 Getter 만으로는 되지 않는 문제가 있었습니다.

JinTaekLim replied:
Data 는 불필요한 어노테이션도 포함이 되기에 @Setter, @Getter 만 활용하는 건 어떨까요?

JinTaekLim commented on src/main/java/com/baedal/gateway/filter/PostAuthenticationFilterFactory.java:
```diff
@@ -0,0 +1,75 @@
+      }
+
+      if (exchange.getResponse().getStatusCode() != HttpStatus.OK) {
+        log.debug("로그인 실패 {}", exchange.getResponse().getStatusCode());
```
exception 이 아닌 로그로 처리하신 이유가 있으실까요?

handsone-u replied:
WebFlux에는 필터에서 발생한 예외를 어떻게 처리할지 모르겠어서 이슈로 등록해두겠습니다.

handsone-u replied:
#8

JinTaekLim replied:
관련 내용은 저도 찾아봐야할 것같아서! 혹여나 먼저 발견하시게 되면 공유 주시면 감사하겠습니다!

handsone-u replied:
https://jaimemin.tistory.com/2583
ErrorWebExceptionHandler 이게 MVC의 컨트롤러 어드바이스와 비슷한 역할을 한다고 합니다.

JinTaekLim commented on src/main/java/com/baedal/gateway/filter/PostAuthenticationFilterFactory.java:
전체적으로 코드를 읽기 어려운 것같습니다.

현재는 
로직 -> 검증 -> 로직 -> 검증 
과 같은 순서로 코드가 작성이 되어있는데,
로직 -> 검증
과 같은 순서로 처리를 하거나, 메소드나 클래스를 구분하는게 좋을 것같습니다.

handsone-u replied:
익명 클래스로 만들어진 부분을 클래스로 분리하는게 좋아보입니다.

JinTaekLim commented on src/main/java/com/baedal/gateway/filter/PreAuthorizationFilterFactory.java:
```diff
@@ -0,0 +1,56 @@
+        }
+      }
+
+      log.debug("UNAUTHORIZED authorization={}", authorizationHeader);
```
로그 관련 코드는 로컬에서 테스트할때만 활용하고, repository 에서는 삭제해두는게 좋을 것같습니다.

handsone-u replied:
각 도메인 별 로그인 관련 기능을 완성하면 그렇게 하겠습니다.

JinTaekLim commented on src/main/java/com/baedal/gateway/filter/PreAuthorizationFilterFactory.java:
```diff
@@ -0,0 +1,56 @@
+          //  - /api/domain/** domain 마다의 role 검증.
+          validator.validateToken(token, config.getRole());
+          log.debug("JWT 검증 성공");
+          return chain.filter(exchange);
```
해당 로직이 수행된 이후, chain.filter 로 넘기게 되면 중간 과정 없이 바로 도메인으로 전달이 되는 걸까요?

handsone-u replied:
중간 과정이 어떤걸까요? 
원래는 다음 필터로 가도록 동작합니다. 하지만 현재 등록된 필터 체인에는 위 필터가 마지막입니다.
그래서 필터는 마무리 되고 도메인으로 라우팅 됩니다.

handsone-u committed:
[#4] add: 로그인 응답 수정(필터)

- id만 응답값에 포함.

handsone-u committed:
[#4] add: 로그인 응답 수정

- id만 응답값에 포함.

handsone-u committed:
[#4] del: role 접두사 제거

- 기존에는 gateway 에서 인증/인가를 하려했으나 수정.
- Spring Security ContextHolder.authority=ROLE_CUSTOMER 이런 형태임.
- 인증된 사용자의 role을 다운 스트림 서비스에서 취급하도록 수정.

handsone-u committed:
[#4] feat: JWT 검증 후 헤더에 인증 정보 추가

- 다운 스트림 서비스로 id, role을 HTTP 헤더에 추가하도록 필터 구현.
- X-User-Id, X-User-Role 을 인증된 사용자의 정보가 필요한 각 도메인에서 Spring security와 인증 필터를 구현하여 인증된 사용자의 정보를 애플리케이션에서 알 수 있도록  구현하면 됨.
- #7 #8

handsone-u committed:
[#4] add: 로그인 예외

- #8

JinTaekLim commented on src/main/java/com/baedal/gateway/filter/PreAuthorizationFilterFactory.java:
```diff
@@ -38,17 +47,41 @@ public GatewayFilter apply(JwtFilterConfig config) {
+          validator.validateToken(token);
+
+          ServerHttpRequest mutatedRequest = exchange.getRequest().mutate()
+              .header("X-User-Id", provider.extractId(token).toString())
```
현재는 게이트웨이에서 헤더에 정보를 넣고, 이를 도메인에서 Context Holder 에 넣는 과정을 거치고 있는데,
게이트웨이에서 ContextHolder 에 값을 넣고 이를 도메인으로 넘겨주는 방안은 어떨까요?

handsone-u replied:
[SecurityContextHolder](https://docs.spring.io/spring-security/reference/servlet/authentication/architecture.html#servlet-authentication-securitycontextholder) 스레드나 애플리케이션에서 관리합니다.

context holder를 요청간에 영속화 해주는 [Persisting Authentication](https://docs.spring.io/spring-security/reference/servlet/authentication/persistence.html#nullsecuritycontextrepository) 쪽도 봤지만 세션 인증방식 외에는 필터가 존재만 하고 아무것도 안하네요. 세션 방식이든 토큰 방식이든 ContextHolder가 서비스간 공유되는 것이 아니지 않나요?

handsone-u committed:
[#4] del: Lombok 애노테이션 수정

- Date를 Getter, Setter로 대체함.

JinTaekLim reviewed (commented):
이전 말씀 나눴던 내용대로라면, 필터가 2개가 존재할 필요가 있을까? 라는 의문이 드는 것같습니다.
괜찮으시다면 필터 별 역할과 동작 순서 등도 함께 작성 주시면 감사하겠습니다

handsone-u committed:
[#2] test: 사용자 role 검증은 다운 스트림 서비스에서

- 65ee41c98d0f90f8267efcea8610700f76b94835 에서 수정한 내용 테스트에 반영

handsone-u committed:
[#4] refactor: 인증 필터 로직 분리

handsone-u commented:
> 이전 말씀 나눴던 내용대로라면, 필터가 2개가 존재할 필요가 있을까? 라는 의문이 드는 것같습니다. 괜찮으시다면 필터 별 역할과 동작 순서 등도 함께 작성 주시면 감사하겠습니다

### PostAuthenticationFilterFactory

실행 흐름 단계
1. 요청 수신 및 라우팅
클라이언트 요청이 Gateway에 도착합니다.
라우팅 정보에 따라 다운스트림 서비스(예: /login 처리 서버)로 요청이 전달됩니다.

2. 다운스트림 응답 수신
다운스트림 서비스에서 응답을 반환하면,
ModifyResponseBodyGatewayFilterFactory가 응답을 가로챕니다.

3. rewriteFunction 실행
모든 로직(/login 경로 확인, 응답 상태 확인 등)은 이 시점에 실행됩니다.

```
(exchange, body) -> {
    // 3-1. 요청 경로 확인 (다운스트림 응답 수신 후)
    String path = exchange.getRequest().getURI().getPath();
    if (!path.endsWith("/login")) return Mono.justOrEmpty(body);

    // 3-2. 응답 상태 코드 확인 (다운스트림 응답 수신 후)
    if (exchange.getResponse().getStatusCode() != HttpStatus.OK) {
        return Mono.error(new RuntimeException("로그인 실패"));
    }

    // 3-3. 응답 본문 파싱 및 JWT 헤더 추가
    // ...
}

```

주의: `exchange.getRequest()`는 원본 요청 객체를 참조합니다.
→ 경로 검사는 다운스트림 응답 수신 후에 이루어지지만,
원본 요청의 경로를 기준으로 필터링합니다.

4. 응답 변조 및 클라이언트 전송
rewriteFunction에서 반환된 최종 응답이 클라이언트에게 전송됩니다.

handsone-u committed:
[#4] add: 토큰 검증 예외 추가

- 인자로 오는 token이 비어있을 시 예외발생

handsone-u committed:
[#4] refactor: JWT 검증 필터를 전역 필터로 변경

- 기존에 있던 PostAuthenticationFilter를 JwtFilter로 변경

handsone-u committed:
[#4] refactor: 라우팅 경로 하드 코딩 제거

handsone-u commented on src/main/java/com/baedal/gateway/filter/JwtFilter.java:
#6

handsone-u committed:
[#8] add: Webflux 전역 예외 핸들러 추가
